File: Java/JDI/jdi-uitest-mobile/src/main/java/com/epam/jdi/uitests/mobile/appium/elements/complex/Selector.java
package com.epam.jdi.uitests.mobile.appium.elements.complex;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.jdi.uitests.core.interfaces.complex.ISelector;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.List;

import static com.epam.commons.EnumUtils.getEnumValue;
import static com.epam.commons.LinqUtils.first;
import static com.epam.commons.LinqUtils.firstIndex;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

/**
 * Created by roman.i on 03.10.2014.
 */

public class Selector<TEnum extends Enum> extends BaseSelector<TEnum> implements ISelector<TEnum> {
    public Selector() {
        super();
    }

    public Selector(By optionsNamesLocatorTemplate) {
        super(optionsNamesLocatorTemplate);
    }

    public Selector(By optionsNamesLocatorTemplate, By allOptionsNamesLocator) {
        super(optionsNamesLocatorTemplate, allOptionsNamesLocator);
    }

    public final void select(String name) {
        actions.select(name, this::selectAction);
    }

    public final void select(TEnum name) {
        select(getEnumValue(name));
    }

    public final void select(int index) {
        actions.select(index, this::selectAction);
    }

    public final String getSelected() {
        return actions.getSelected(this::getSelectedAction);
    }

    public final int getSelectedIndex() {
        return actions.getSelectedIndex(this::getSelectedIndexAction);
    }

    protected final boolean isSelectedAction(String name) {
        return getSelectedAction().equals(name);
    }

    protected final boolean isSelectedAction(int index) {
        return getSelectedIndexAction() == index;
    }

    protected String getValueAction() {
        return getSelected();
    }

    protected String getSelectedAction() {
        if (allLabels() != null)
            return getSelected(allLabels().getWebElements());
        if (getLocator().toString().contains("%s"))
            throw exception("Can't get Selected options. Override getSelectedAction or place locator to <select> tag");
        List<WebElement> els = getAvatar().searchAll().getElements();
        if (els.size() == 1)
            return getSelected(getSelector().getOptions());
        else
            return getSelected(els);
    }

    private String getSelected(List<WebElement> els) {
        WebElement element = first(els, this::isSelectedAction);
        if (element == null)
            throw exception("No elements selected. Override getSelectedAction or place locator to <select> tag");
        return element.getText();
    }

    protected int getSelectedIndexAction() {
        if (allLabels() != null) {
            return getSelectedIndex(allLabels().getWebElements());
        }
        if (getLocator().toString().contains("%s"))
            throw exception("Can't get Selected options. Override getSelectedAction or place locator to <select> tag");
        List<WebElement> els = getAvatar().searchAll().getElements();
        if (els.size() == 1)
            return getSelectedIndex(getSelector().getOptions());
        else
            return getSelectedIndex(els);
    }

    private int getSelectedIndex(List<WebElement> els) {
        int index = firstIndex(els, this::isSelectedAction) + 1;
        if (index == 0)
            throw exception("No elements selected. Override getSelectedAction or place locator to <select> tag");
        return index;
    }
}

File: Python/JDI/web/selenium/elements/complex/selector.py
from enum import Enum

from JDI.core.utils.decorators import scenario
from JDI.web.selenium.elements.complex.base_selector import BaseSelector


class Selector(BaseSelector):
    def __init__(self, by_option_locator_template=None, by_option_locator_all=None):
        if by_option_locator_all is None:
            super(Selector, self).__init__(by_option_locator_template=by_option_locator_template)
        else:
            super(Selector, self).__init__(by_option_locator_template=by_option_locator_template,
                                           by_option_locator_all=by_option_locator_all)

    def is_selected_action(self, el):
        if isinstance(el, str):
            return self.get_selected() == el
        if isinstance(el, Enum):
            return self.get_selected() == el.value
        else:
            super(Selector, self).is_selected(el)

    def get_names(self):
        return self.get_options()

    def get_options_as_text(self):
        return ", ".join(self.get_options())

    @scenario(action_name="Get Selected element name")
    def get_selected(self):
        return self.get_selected_action()

    def get_selected_action(self):
        el = list(filter(lambda x: x.is_selected(), self.get_web_elements()))
        if len(el) == 0:
            raise Exception("No elements selected. Override getSelectedAction or place locator to <select> tag")
        return el[0].text

    @scenario(action_name="Get Selected element index")
    def get_selected_index(self):
        try:
            index = self.get_selected_index_action()
            if index is None:
                raise ValueError("No elements selected. Override getSelectedAction or place locator to <select> tag")
            else:
                return index
        except ValueError:
            raise ValueError("No elements selected. Override getSelectedAction or place locator to <select> tag")

    def get_selected_index_action(self):
        return list(map(lambda x: x.is_selected(), self.get_web_elements())).index(True)

    def get_values(self):
        return self.get_options()

    def get_value_action(self):
        return self.get_selected()

    @scenario(action_name="Select")
    def select(self, name):
        self.select_action(name)
